Add Flutter SDK migration guide (#239)
amplitude/amplitude-dev-center: docs/data/sdks/flutter/migration.md

---
title: Flutter SDK Migration Guide
description: Use this guide to easily migrate from Amplitude's Flutter SDK 3.0 to the Flutter SDK 4.0.
---

> Amplitude Flutter SDK 4.0 features default event tracking, simplified interfaces and wraps on the latest [Amplitude iOS SDK](/data/sdks/ios-swift/) and [Android Kotlin SDK](/data/sdks/android-kotlin/). 

Reviewer: ```suggestion
Amplitude Flutter SDK 4.0 features default event tracking, simplified interfaces, and wraps the latest [Amplitude iOS](/data/sdks/ios-swift/) and [Android Kotlin](/data/sdks/android-kotlin/) SDKs. 
```

!!!beta "SDK Resources"
     [:material-github: GitHub](https://github.com/amplitude/Amplitude-Flutter/tree/beta) · [:material-code-tags-check: Releases](https://github.com/amplitude/experiment-php-server/releases)

## Terminology

* `amplitude_flutter:v3`: Flutter SDK 3.0
* `amplitude_flutter:v4`: Flutter SDK 4.0

## Dependencies

> Go to the pubspec.yaml file and update the dependency.

Reviewer: ```suggestion
Open `pubspec.yaml` and update the dependency:
```

```diff
dependencies:
-  amplitude_flutter: ^3
+  amplitude_flutter: ^4
```

## Instrumentation

> Flutter SDK 4.0 offers an API to instrument events. To migrate to the Flutter SDK 4.0, you need to update a few calls. The following sections detail which calls have changed.

Reviewer: ```suggestion
Flutter SDK 4.0 offers an API to instrument events. To migrate to Flutter SDK 4.0, you need to update a few calls. The following sections detail which calls have changed.
```

### Initialization

> Like all other calls, `instance()` has been removed. Configuration is handled differently. The Flutter SDK 4.0 use the Configuration object to set the configuration. See [Configuration](#configuration).

Reviewer: ```suggestion
Like all other calls, `instance()` is removed. Flutter SDK 4.0 uses the Configuration object to set the configuration. See [Configuration](#configuration) for more information.
```

```diff
- import 'package:amplitude_flutter/amplitude.dart';
- import 'package:amplitude_flutter/identify.dart';
+ import 'package:amplitude_flutter/amplitude.dart';
+ import 'package:amplitude_flutter/configuration.dart';

// Create the instance and initialize SDK
- final Amplitude amplitude = Amplitude.getInstance(instanceName: "project");
- amplitude.setServerUrl("https://your.endpoint.com")
- amplitude.setServerZone("US")
- amplitude.trackingSessionEvents(true)
- amplitude.init(widget.apiKey);
+ final Amplitude amplitude = Amplitude(Configuration(
+     apiKey: "YOUR-API-KEY",
+     serverUrl: "https://your.endpoint.com",
+     serverZone: ServerZone.eu,
+     defaultTracking: DefaultTrackingOptions(
+        sessions: true
+     )
+   ));
+ await amplitude.isBuilt;
```

### Configuration

> The configurations for the new SDK are simpler and more consistent across runtimes. Some configurations are no longer supported. The Flutter SDK 3.0 uses setters for configuration. The configuration of the Flutter SDK 4.0 can be set by an instance variable. Note the Configuration is immutable (cannot be changed) after being passed to Amplitude.

Reviewer: ```suggestion
Flutter SDK 4.0 includes the following configuration changes:

- It's more consistent across runtimes.
- It no longer supports certain configurations.
- It uses an instance variable for configuration, versus setters in SDK version 3.0.
- Configuration is immutable after you pass it to Amplitude.
```

|Flutter SDK 3.0                      |Flutter SDK 4.0                     |
|-------------------------------------|------------------------------------|
| `enableCoppaControl()` or `disableCoppaControl()` | `config.enableCoppaControl`         |
| `setMinTimeBetweenSessionsMillis()` | `config.minTimeBetweenSessionsMillis` |
| `setEventUploadThreshol(d)`         | `config.flushQueueSize`            |
| `setEventUploadPeriodMillis()`      | `config.flushIntervalMillis`       |
| `setServerZone()`                   | `config.serverZone`                |
| `setServerUrl()`                    | `config.serverUrl`                 |
| `setUseDynamicConfig()`             | NOT SUPPORTED                      |
| `setOptOut()`                       | `config.optOut`                    |
| `setOffline()`                      | NOT SUPPORTED                      |
| `trackingSessionEvents()`           | `config.defaultTracking.sessions`  |
| `useAppSetIdForDeviceId()`          | `config.useAppSetIdForDeviceId`    |

### Tracking events

> The Flutter SDK 3.0 offered a variety of `logEvent` APIs with `withEventProperties`, `withApiProperties`, `withUserProperties`, `withGroup`, `withGroupProperties`, `withTimestamp`, `outOfSession`, to override specific properties in the event payload. Amplitude has simplified all these variations into a unified `track` API.

Reviewer: ```suggestion
Flutter SDK 4.0 uses a unified `track` API to replace the following `logEvent` API variations:

- `withEventProperties`
- `withApiProperties`
- `withUserProperties`
- `withGroup`
- `withGroupProperties`
```

Changed the order of reference here and made it a list.

| `enableCoppaControl()` or `disableCoppaControl()` | `config.enableCoppaControl`         |
| `setMinTimeBetweenSessionsMillis()` | `config.minTimeBetweenSessionsMillis` |
| `setEventUploadThreshol(d)`         | `config.flushQueueSize`            |
| `setEventUploadPeriodMillis()`      | `config.flushIntervalMillis`       |
| `setServerZone()`                   | `config.serverZone`                |
| `setServerUrl()`                    | `config.serverUrl`                 |
| `setUseDynamicConfig()`             | NOT SUPPORTED                      |
| `setOptOut()`                       | `config.optOut`                    |
| `setOffline()`                      | NOT SUPPORTED                      |
| `trackingSessionEvents()`           | `config.defaultTracking.sessions`  |
| `useAppSetIdForDeviceId()`          | `config.useAppSetIdForDeviceId`    |

### Tracking events

Flutter SDK 4.0 uses a unified `track` API to replace the following `logEvent` API variations:

* `withEventProperties`
* `withApiProperties`
* `withUserProperties`
* `withGroup`
* `withGroupProperties`

#### `logEvent()`

The `logEvent()` API maps to `track()`.

```diff
- amplitude.logEvent('BUTTON_CLICKED');

> + amplitude.track(event: BaseEvent(eventType:'BUTTON_CLICKED'))

Reviewer 2: nit: this usage seems verbose. If we always need to pass an event, can we make the parameter unnamed?

`amplitude.track(BaseEvent(eventType:'BUTTON_CLICKED'))`

Author: I noticed that two after the release. I've put it to the backlog and will update it in the next beta version. https://amplitude.atlassian.net/wiki/spaces/GOV/pages/2255355925/Flutter+SDK+update+to+wrap+on+the+latest+SDKs#Backlog

```diff
- amplitude.uploadEvents();
+ amplitude.flush()
```

### Set user properties

The APIs for setting user properties are the same, except for the removal of `instance()`. Here are code snippets to migrate APIs for user properties.

#### `setUserId()`

`setUserId()` remains unchanged but does **NOT** receive `startNewSession`.

#### `setDeviceId()`

`setDeviceId()` remains unchanged.

#### `regenerateDeviceId()`

`regenerateDeviceId()` is not supported in the Flutter SDK 4.0.

#### `clearUserProperties()`

The `clearUserProperties` API has been removed, but you can now use the unified `identify` API to remove user properties. 

```diff
- amplitude.clearUserProperties();
+ final Identify identify = Identify()
+   ..clearAll()
+ amplitude.identify(identify)
```

#### `setUserProperties()`

The `setUserProperties` API has been removed, but you can now use the unified `identify` API to add user properties. 

```diff
- Map<String, dynamic> userProps = {
-   'gender': 'female',
-   'age': '20'
- };
- amplitude.setUserProperties(userProperties);
+ final Identify identify = Identify()
+                          ..set('gender','female')
+                          ..set('age',20);

> + amplitude.identify(identify);

Author: `;` is required. line 161, 162, and 163 are the same line of code. See https://dart.dev/language/operators#cascade-notation.

> + amplitude.identify(identify);

Reviewer 3: I saw other examples do not have `;`, could you add all?

> + amplitude.identify(identify);

Author: Added at https://github.com/amplitude/amplitude-dev-center/pull/239/commits/a11189de6e73b34fea2b84edcc6a8e14da5fdc5c
